pkg/cmd/secrets.go

```diff
@@ -380,6 +408,89 @@ func downloadSecrets(cmd *cobra.Command, args []string) {
 	utils.Log(fmt.Sprintf("Downloaded secrets to %s", filePath))
 }
 
+func templateSecrets(cmd *cobra.Command, args []string) {
+	localConfig := configuration.LocalConfig(cmd)
+
+	utils.RequireValue("token", localConfig.Token.Value)
+
+	filePath, err := utils.GetFilePath(args[0])
+	if err != nil {
+		utils.HandleError(err, "Unable to parse template file path")
+	}
+
+	var file []byte
+	file, err = ioutil.ReadFile(filePath)
+	if err != nil {
+		utils.HandleError(err, "Unable to read template file")
+	}
+
+	var outputFile *os.File
+	var outputFilePath string
+	output := cmd.Flag("output").Value.String()
+	if len(output) != 0 {
+		outputFilePath, err = utils.GetFilePath(output)
+		if err != nil {
+			utils.HandleError(err, "Unable to parse output file path")
+		}
+		outputFile, err = os.Create(outputFilePath)
+		if err != nil {
+			utils.HandleError(err, "Unable to open output file for writing")
+		}
+		defer outputFile.Close()
+	}
+
+	funcs := map[string]interface{}{
+		"tojson": func(value string) (string, error) {
+			body, err := json.Marshal(value)
+			if err != nil {
+				return "", err
+			}
+			return string(body), nil
+		},
+		"fromjson": func(value string) (interface{}, error) {
+			var result interface{}
+			err = json.Unmarshal([]byte(value), &result)
+			if err != nil {
+				return "", err
+			}
+			return result, nil
+		},
+	}
+	template, err := template.New("Secrets").Funcs(funcs).Parse(string(file))
+	if err != nil {
+		utils.HandleError(err, "Unable to parse template text")
+	}
+
+	response, responseErr := http.GetSecrets(localConfig.APIHost.Value, utils.GetBool(localConfig.VerifyTLS.Value, true), localConfig.Token.Value, localConfig.EnclaveProject.Value, localConfig.EnclaveConfig.Value)
+	if !responseErr.IsNil() {
+		utils.HandleError(responseErr.Unwrap(), responseErr.Message)
+	}
+
+	secrets, parseErr := models.ParseSecrets(response)
+	if parseErr != nil {
+		utils.HandleError(parseErr, "Unable to parse API response")
+	}
+
+	secretsMap := map[string]interface{}{}
+	for _, secret := range secrets {
+		secretsMap[secret.Name] = secret.ComputedValue
+	}
+
+	if outputFile != nil {
+		err = template.Execute(outputFile, secretsMap)
```

Comment:
I think this command should be atomic- we either write the completed template, or we write nothing. It seems this is counter to how `Execute` works though: `If an error occurs executing the template or writing its output, execution stops, but partial results may already have been written to the output writer.`

Do you have any ideas how we can get around this? I'm thinking maybe we write to a local writer and then if that succeeds we write the contents to the specified file/stdout. This would also solve for the scenario I mentioned above, where the stdout output was confusing.

Reply:
Yeah, this is a little weird for sure. It seems like it might be done this way to allow the templating to work on very large files that wouldn't otherwise fit into memory. I'd prefer to optimize for user experience on smaller files and follow your local-writer suggestion.

Reply:
Tested this new functionality and it works exactly as expected. Nice job!